### PR TITLE
JDK-8318671: Potential uninitialized uintx value after JDK-8317683

### PR DESCRIPTION
**Pull request by tstuefe**
When using 'MemStat' CompileCommand, we accidentally register the command if an invalid suboption had been specified. Fixed, added regression test (verified).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318671](https://bugs.openjdk.org/browse/JDK-8318671): Potential uninitialized uintx value after JDK-8317683 (**Bug** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [287d8106](https://git.openjdk.org/jdk/pull/16335/files/287d8106c3d6c273ade3bbe2ccd93c8d924d8897)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16335/head:pull/16335` \
`$ git checkout pull/16335`

Update a local copy of the PR: \
`$ git checkout pull/16335` \
`$ git pull https://git.openjdk.org/jdk.git pull/16335/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16335`

View PR using the GUI difftool: \
`$ git pr show -t 16335`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16335.diff">https://git.openjdk.org/jdk/pull/16335.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16335#issuecomment-1776806645)

### CODE REVIEW

File: src/hotspot/share/compiler/compilerOracle.cpp
Comment (tstuefe) [reply src/hotspot/share/compiler/compilerOracle.cpp:650]: > If "omitting value means 'collect'" then why do we not simply set value = (uintx)MemStatAction::collect ? Otherwise what is that message supposed to mean?

We only enter this function if a value for this command had been given (`-XX:CompileCommand=<command>,<method>[,<value>]`). The default case is handled somewhere else (CompilerOracle::parse_from_line).

> Ternary return values are unpleasant.

As, on a general base? You don't like the number three :-) ?

Here, we have to distinguish between:
- not handled, its not a command option that accepts enum strings
- handled, with two variants: handled, had an error, and handled, had no error. Only in the latter case we want to register the command.

Comment (dholmes-ora) [reply src/hotspot/share/compiler/compilerOracle.cpp:650]: Okay I'll butt out. The comment doesn't make sense to me.

File: test/hotspot/jtreg/serviceability/dcmd/compiler/CompilerMemoryStatisticTest.java
Comment (tstuefe) [inline test/hotspot/jtreg/serviceability/dcmd/compiler/CompilerMemoryStatisticTest.java:53]: Reviewer hint: Removed because its redundant with the syntax check introduced in MemStatTest.java

File: src/hotspot/share/compiler/compilerOracle.cpp
Comment (tstuefe) [reply src/hotspot/share/compiler/compilerOracle.cpp]: @dholmes-ora Sorry if I came across as flippant; that was not my intention. If you can stomach it, I rewrote the patch to be hopefully easier to read. Thanks.

Comment (dholmes-ora) [reply src/hotspot/share/compiler/compilerOracle.cpp]: No I didn't take it that way at all. The new patch does seem easier to read, but I will still defer to compiler folk to review this.

File: test/hotspot/jtreg/compiler/compilercontrol/commands/MemStatTest.java
Comment (jdksjolen) [inline test/hotspot/jtreg/compiler/compilercontrol/commands/MemStatTest.java:3]: Thanks! But I think we're OK with Redhat taking credit for this one :).

File: src/hotspot/share/compiler/compilerOracle.cpp
Comment (jdksjolen) [reply src/hotspot/share/compiler/compilerOracle.cpp:654]: Ouch, I just realized that we can't differentiate between being provided with the literal number 2 and `MemStatAction::print` anymore since you moved the literal number parsing into this function. That means that we can't set `print_final_memstat_report` after returning from this function.

**Review by jdksjolen — COMMENTED** [review]
Hi,

This looks good to me, I've got a couple of comments (one is just a nit). I'll leave it to the compiler folks for approval.

File: src/hotspot/share/compiler/compilerOracle.cpp
Comment (tstuefe) [reply src/hotspot/share/compiler/compilerOracle.cpp:654]: > Ouch, I just realized that we can't differentiate between being provided with the literal number 2 and `MemStatAction::print` anymore since you moved the literal number parsing into this function. That means that we can't set `print_final_memstat_report` after returning from this function.

Yes, its a big annoyance with unnecessary  templatization in this file. The natural point of "print_final.." would be in register_command, but I cannot (easily) query values there because value is T. So I cannot just do a "if type is uintx and option is Memstat and value is print then do xxx".

This code would gain IMHO by being dumbed down. A multi-type value can just as well be expressed with a union, no templates needed. Especially since we keep the type info around at runtime anyway.

File: test/hotspot/jtreg/compiler/compilercontrol/commands/MemStatTest.java
Comment (tstuefe) [reply test/hotspot/jtreg/compiler/compilercontrol/commands/MemStatTest.java:3]: Oops :-)

File: src/hotspot/share/compiler/compilerOracle.cpp
Comment (jdksjolen) [reply src/hotspot/share/compiler/compilerOracle.cpp:654]: >This code would gain IMHO by being dumbed down

Agreed, seems like it's had a lot of "just one more feature" commits over time.

**Review by TobiHartmann — APPROVED** [review]
Looks reasonable to me.

**Comment by TobiHartmann** [comment]
I think this is ready for integration given that both @dholmes-ora and @jdksjolen are okay with it.

**Comment by tstuefe** [comment]
> I think this is ready for integration given that both @dholmes-ora and @jdksjolen are okay with it.

Well, they did not approve yet; @jdksjolen or @dholmes-ora, if you are happy with this, could you hit the big green button please?

**Comment by dholmes-ora** [comment]
@TobiHartmann both @jdksjolen  and myself indicated we would defer to the compiler team to review/approve this. I only had some drive-by comments.

File: src/hotspot/share/compiler/compilerOracle.cpp
Comment (shipilev) [inline src/hotspot/share/compiler/compilerOracle.cpp:678]: It is honestly weird to see `parse***Uintx***Value` dealing with enums, and be specialized for `MemStat`. Can you reflow this to match how `MemLimit` does it? https://github.com/openjdk/jdk/blob/7bb1999c51cdfeb020047e1094229fda1ec5a99d/src/hotspot/share/compiler/compilerOracle.cpp#L702

Comment (tstuefe) [reply src/hotspot/share/compiler/compilerOracle.cpp]: Okay, rewritten in the style of https://github.com/openjdk/jdk/pull/16631. Retested too. Let's get this out of the door, please.

**Review by shipilev — APPROVED** [review]
All right, new version looks significantly better, thanks!

**Comment by tstuefe** [comment]
MacOS error unrelated.

/integrate

**Pull request closed without merging**
